crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,317 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/buffer.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/mem.h>
+#include <openssl/pkcs7.h>
+#include <stdio.h>
+#include "../../fipsmodule/cipher/internal.h"
+#include "../internal.h"
+
+static int enc_write(BIO *h, const char *buf, int num);
+static int enc_read(BIO *h, char *buf, int size);
+static long enc_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int enc_new(BIO *h);
+static int enc_free(BIO *data);
+#define ENC_BLOCK_SIZE (1024 * 4)
+
+typedef struct enc_struct {
+  uint8_t done;     // > 0 when finished
+  uint8_t flushed;  // > 0 when buffered data has been flushed
+  uint8_t ok;       // decryption status
+  int buf_off;      // start idx of buffered data
+  int buf_len;      // length of buffered data
+  EVP_CIPHER_CTX *cipher;
+  uint8_t buf[ENC_BLOCK_SIZE];
+} BIO_ENC_CTX;
+
+static const BIO_METHOD methods_enc = {
+    BIO_TYPE_CIPHER,
+    "cipher",
+    enc_write,
+    enc_read,
+    NULL,  // enc_puts
+    NULL,  // enc_gets
+    enc_ctrl,
+    enc_new,
+    enc_free,
+    NULL,  // enc_callback_ctrl
+};
+
+const BIO_METHOD *BIO_f_cipher(void) { return &methods_enc; }
+
+static int enc_new(BIO *b) {
+  BIO_ENC_CTX *ctx;
+  GUARD_PTR(b);
+
+  if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL) {
+    OPENSSL_PUT_ERROR(BIO, ERR_R_MALLOC_FAILURE);
```

Comment:
Just in case this comment gets lost: https://github.com/aws/aws-lc/pull/1836#discussion_r1779286649